openedx/core/djangoapps/course_roles/data.py

```diff
@@ -0,0 +1,160 @@
+"""
+Permissions for course roles app.
+
+These are the permissions that can be assigned to a CourseRole which grants access for a course,
+but can also be assigned for org wide course access or instance wide course access.
+They are defined in the database in the course_roles_permission table.
+
+To add a new permission, add a new entry to the CourseRolesPermission enum,
+then and add a new row to the course_roles_permission table in database,
+with a migration.
+
+To remove a permission, remove the entry from the CourseRolesPermission enum,
+and remove the row from the course_roles_permission table in database,
+with a migration.
+
+To change the readable_name or description of a permission, change the
+corresponding entry in CourseRolesPermission enum.
+
+To change the name of a permission, change the corresponding entry on the
+CourseRolesPermission enum, and change the name field of the corresponding
+row in the course_roles_permission table in database, with a migration.
+"""
+from attrs import frozen, field, validators
+from enum import Enum, unique
+
+from django.utils.translation import gettext as _
+
+
+@frozen
+class PermissionData:
+    """
+    Data class for a permission.
+    """
+    name: str = field(validator=validators.instance_of(str))
+    readable_name: str = field(validator=validators.instance_of(str))
+    description: str = field(validator=validators.instance_of(str))
+
+
+@unique
+class CourseRolesPermission(Enum):
+    """
+    Enum of all user permissions, the values are the permissions names
+    in the course_roles_permission table in database.
+
+    The readable_name and description are used in the UI.
+    """
```

Comment:
Long term note: One architectural failure mode I'm worried about here is if people find it difficult to add new permissions and instead start using other permissions that imply a role.

For instance, say someone wants to add a permission for the Staff Debug information under Components in the LMS rendering. But adding a permission and doing the data migration sounds hard. Then they browse through this class and think to themselves, "Well, if they can `VIEW_ALL_CONTENT` AND `MANAGE_STUDENTS`, then..." and we get into the mode where people are re-deriving roles implicitly from permissions and extrapolating those onto other implied permissions.

I'm not saying we're doomed to this. Just that it's a risk and that we'll need docs and outreach, presentations, etc. in order to make sure that developers working on this will have the right mindset. This gets more likely if there's any real friction to adding permissions–long approval process, unclear docs, any perception of operational risk in the event of a rollback, etc.

Come to think on it, this whole thing would make for a great Open edX Conference talk or a Meetup presentation.

Reply:
We agree this presents a risk and have been thinking through strategies to mitigate for this risk. We do not think this presents an immediate risk, but plan to add documentation around how and when to add a permission in [OEP-66](https://open-edx-proposals.readthedocs.io/en/latest/best-practices/oep-0066-bp-authorization.html) and within the djangoapps/course_roles/docs folder.